keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/VersioningStrategy.kt

```diff
@@ -0,0 +1,148 @@
+/*
+ *
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.keel.api
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.databind.JsonDeserializer
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.netflix.frigga.ami.AppVersion
+import com.netflix.rocket.semver.shaded.DebianVersionComparator
+import com.netflix.spinnaker.keel.api.SortType.INCREASING
+import com.netflix.spinnaker.keel.api.SortType.SEMVER
+import com.netflix.spinnaker.keel.serialization.PropertyNamePolymorphicDeserializer
+import net.swiftzer.semver.SemVer
+import org.slf4j.LoggerFactory
+import org.springframework.util.comparator.NullSafeComparator
+
```

Comment:
I moved all the docker sorting logic to be in proper comparators, so that it was easy to sort lists with them.